### ISSUE TITLE
LXC's seccomp filter v2 format parser generates incomplete BPF programs

### ISSUE THREAD

**Issue by minipli** [lxc/lxc#571]
Trying to handle the case where the syscall number for the native architecture matches the one of the compat architecture is wrong. The syscall shouldn't be added to the native arch context only but to all contexts in any case. libseccomp can handle this case just fine. LXC's current behavior with the following seccomp filter leads to this:

> ,-[ lxc.seccomp ]-
> | 2
> | blacklist
> | [x86]
> | syslog errno 1
> `---

The above filter will generate to a BPF program that filters the `syslog` syscall for **x86-64** while it should filter it for **x86** instead, just because the syscall numbers are identical for both ABIs. This can be confirmed when dumping the generated BPF program (using libseccomp's `scmp_bpf_disasm`):

```
 line  OP   JT   JF   K
=================================
 0000: 0x20 0x00 0x00 0x00000004   ld  $data[4]
 0001: 0x15 0x00 0x04 0xc000003e   jeq 3221225534 true:0002 false:0006
 0002: 0x20 0x00 0x00 0x00000000   ld  $data[0]
 0003: 0x35 0x02 0x00 0x40000000   jge 1073741824 true:0006 false:0004
 0004: 0x15 0x00 0x02 0x00000067   jeq 103  true:0005 false:0007
 0005: 0x06 0x00 0x00 0x00050001   ret ERRNO(1)
 0006: 0x15 0x00 0x01 0x40000003   jeq 1073741827 true:0007 false:0008
 0007: 0x06 0x00 0x00 0x7fff0000   ret ALLOW
 0008: 0x06 0x00 0x00 0x00000000   ret KILL
```

The x86 specific seccomp filter at `0006` is empty. It directly branches to the `ret ALLOW` instruction at `0007` which mirrors the default policy. OTOH, the x86-64 specific filter starting at `0004` tests the syscall number to be 103 (`syscall`) and branches to `ret ERRNO(1)` if it matches. So exactly the opposite than what the filter actually should do.

The very same bug also leads to handling the special rule `reject_force_umount` incorrectly as can be seen below

> ,-[ lxc.seccomp ]-
> | 2
> | blacklist
> | [all]
> | reject_force_umount
> `---

The above filter leads to a BPF program that filters `umount2(…,MNT_FORCE)` for x86-64 only - not for the compat architecture. This allows an x86 program running on a x86-64 kernel confined with that seccomp policy to still issue forced umounts. The reason is, again, the special casing for the syscall numbers match for native and compat arches. But in this case the syscall will be `-1` as there is no syscall named `reject_force_umount`.

Here is the generated BPF program:

```
 line  OP   JT   JF   K
=================================
 0000: 0x20 0x00 0x00 0x00000004   ld  $data[4]
 0001: 0x15 0x00 0x09 0xc000003e   jeq 3221225534 true:0002 false:0011
 0002: 0x20 0x00 0x00 0x00000000   ld  $data[0]
 0003: 0x35 0x07 0x00 0x40000000   jge 1073741824 true:0011 false:0004
 0004: 0x15 0x00 0x07 0x000000a6   jeq 166  true:0005 false:0012
 0005: 0x20 0x00 0x00 0x0000001c   ld  $data[28]
 0006: 0x15 0x00 0x05 0x00000000   jeq 0    true:0007 false:0012
 0007: 0x20 0x00 0x00 0x00000018   ld  $data[24]
 0008: 0x54 0x00 0x00 0x00000001   and 0x00000001
 0009: 0x15 0x00 0x02 0x00000001   jeq 1    true:0010 false:0012
 0010: 0x06 0x00 0x00 0x0005000d   ret ERRNO(13)
 0011: 0x15 0x00 0x01 0x40000003   jeq 1073741827 true:0012 false:0013
 0012: 0x06 0x00 0x00 0x7fff0000   ret ALLOW
 0013: 0x06 0x00 0x00 0x00000000   ret KILL
```

The x86 specific path through the program (`0001` -> `0011` -> `0012`) directly leads to the `ret ALLOW` instruction without any syscall checks. Again, not intended by the original seccomp policy.

**Comment by minipli** [comment]
Just a short update. The seccomp filter for `reject_force_umount` can be bypassed on x86-64 as well, due to a bug in libsseccomp. One might see it from the BPF program above. The test in line `0006` for zero in the upper 32 bit of the second argument is wrong. It's not needed but allows branching to the default policy (ALLOW) in case it's not zero. The kernel OTOH ignores what's in the upper 32 bit. So the following code snippet bypasses the seccomp filter on x86-64 and successfully calls `umount2()` with the `MNT_FORCE` flag set:
```
#define _GNU_SOURCE
#include <sys/syscall.h>
#include <sys/mount.h>
#include <unistd.h>

int main(int argc, char **argv) {
    if (argc != 2)
        return 1;

    return syscall(SYS_umount2, argv[1], MNT_FORCE | (1UL << 32));
}
```

The libseccomp bug is not fixed, yet. But a patch has been posted [here](https://groups.google.com/d/msg/libseccomp/VtrClkXxLGA/ZsuJ5uQi8AYJ).

**Comment by stgraber** [comment]
Hi,

So first of all, we're aware that our seccomp profile is bypassable and not just for the mount stuff. Any seccomp profile which doesn't block ptrace can be bypassed by the user through the use of ptrace (it's a documented limitation of seccomp).

That's pretty unfortunate as that means we can't make privileged containers as safe as we'd like, though at the same time we've always stated that people who are handing container to untrusted users really should be using unprivileged containers.

There are a few patches floating around. One by Eric Biedermann to fix the specific fuse issue which caused us to block force umount and another by Andy Lutomirski which attempts to solve the ptrace limitation for seccomp.

**Comment by minipli** [comment]
I'm aware of the ptrace bypass. That hole can easily be plugged by providing a seccomp policy which includes `ptrace`. Well, maybe not in a "common seccomp profile", as that would prevent using strace, gdb and friends. But still, in a user provided policy it would be possible.

I'm also aware that LXC is preventing libseccomp from issuing the `pctrl(PR_SET_NO_NEW_PRIVS, 1)`. That cannot by "fixed" by a user as LXC provides no knob to do so.

Btw., `pctrl(PR_SET_NO_NEW_PRIVS, 1)` would not only prevent raising the capabilities beyond what's currently set but at the same time would disable abusing the seccomp filter to do harm, i.e. with seccomp policies like...
```
...
setuid errno 0
getuid errno 0
```
...to prevent SUID programs to drop there privileges. So, why does LXC disable NNP in libseccomp, i.e. from doing "the right thing" in the first place? Oh, I know -- because of `ping`.

But still. This is all no excuse for doing it wrong in the first place. LXC's usage of libseccomp is broken -- simply broken, nothing else. So can we fix that please?

I provided two seccomp policy files that both generate wrong BPF programs. Not only just because of bugs in libsecomp but because of bugs in LXC. Aren't those worth fixing?

**Comment by hallyn** [comment]
Patches to provide a switch for NNP are welcome.

Regarding the seccomp profiles, I'm not finding that to be the case.  If I add

[x86_64]
read errno 1

to a x86-64 container's profile, it cannot read.  When I add

[x86]
read errno 1

it can.  That looks correct.  Likewise, when I do

[all]
reject_force_umount  # comment this to allow umount -f;  not recommended

I indeed cannot umount -f, as is expected.

**Comment by hallyn** [comment]
Note that reject_force_umount is a special case keyword, which ignores
'[all]' and '[x86]'.

**Comment by minipli** [comment]
So you must be using a different version of LXC or libseccomp than I am. `umount -f` works like a charm, here:

```
root@marvin:/chroot# lxc-execute -n Foo -f lxc.conf /bin/bash
root@lxc:/# ls -ld /proc/1/
dr-xr-xr-x 9 root root 0 Jun 25 12:49 /proc/1/
root@lxc:/# umount -f /proc
root@lxc:/# ls -ld /proc/1/
ls: cannot access /proc/1/: No such file or directory
root@lxc:/# file $(which umount) | sed 's, dyn.*,...,'
/bin/umount: setuid ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV),...
root@lxc:/# uname -m
x86_64
root@lxc:/# exit
root@marvin:/chroot# cat lxc.conf
lxc.utsname = lxc
lxc.seccomp = lxc.seccomp
lxc.rootfs = /chroot/wheezy-i386
root@marvin:/chroot# cat lxc.seccomp
2
blacklist
[all]
reject_force_umount
root@marvin:/chroot#
```

These are the versions I'm using:

```
minipli@marvin:~$ apt-cache show libseccomp2 | grep -i version
Version: 2.1.1-1
minipli@marvin:~$ cd src/lxc/
minipli@marvin:~/src/lxc (master)$ git describe HEAD
lxc-1.1.0-177-gd791668b9adf
```

The version of libseccomp is slightly outdated, admitted. But I also tried using the latest git version with no different results.

So, I'm very well aware that `reject_force_umount` is a special keyword. I even mentioned it in the initial bug report, didn't I? But apparently the code does not, as it tries to resolve `reject_force_umount` to a syscall number which triggers the faulty behavior in the rest of the code.


To respond to your `read` policy test. I didn't say there would be no cases where the seccomp filter would be correct. I instead showed you two cases were the generated filter is incorrect. If you're testing the ones where the code gets it right you won't see faulty behavior. Test the ones I provided instead, please.

This is what I get for the `syslog` policy:

```
root@marvin:/chroot# lxc-execute -n Foo -f lxc.conf /bin/bash
root@lxc:/# /root/dmesg64 | head -n 1
dmesg64: klogctl failed: Operation not permitted
root@lxc:/# /root/dmesg32 | head -n 1
[    0.000000] Initializing cgroup subsys cpuset
root@lxc:/# file /root/dmesg* | sed 's/ dyn.*/.../'
/root/dmesg32: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV),...
/root/dmesg64: ELF 64-bit LSB executable, x86-64, version 1 (SYSV),...
root@lxc:/# uname -m
x86_64
root@lxc:/# exit
root@marvin:/chroot# cat lxc.conf
lxc.utsname = lxc
lxc.seccomp = lxc.seccomp
lxc.rootfs = /chroot/wheezy-amd64
root@marvin:/chroot# cat lxc.seccomp
2
blacklist
[x86]
syslog errno 1
```

As already noted in the initial bug report, exactly the opposite from what I would have expected.

**Comment by minipli** [comment]
Yet another short update. The libseccomp bug mentioned in the second comment has been fixed meanwhile and ended up in version v2.2.2. So the bugs left are all in LXC's code base...

**Comment by brauner** [comment]
This does not seem to be a problem with newer version of `LXC`. Also, NNP can be set with `lxc.no_new_privs = 1` in the containers config file.
